ingress-node-firewall/v1alpha1/ingressnodefirewall_types.go

```diff
@@ -0,0 +1,163 @@
+type IngressNodeFirewallProtoRule struct {
+	//Ports can be a single port or range of ports using start-end range format
+	//+kubebuilder:validation:Optional
+	// +optional
+	Ports string `json:"ports,omitempty"`
+}
+
+// IngressNodeFirewallProtocolRule define ingress node firewall rule per protocol
+type IngressNodeFirewallProtocolRule struct {
+	// Order define order of execution of ingress firewall rules .
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Minimum:=1
+	Order uint32 `json:"order"`
+
+	// IngressNodeFirewallProtoRule define ingress node firewall rule for TCP, UDP and SCTP protocols.
+	// +optional
+	ProtocolRule *IngressNodeFirewallProtoRule `json:"protoRule,omitempty"`
+
+	// IngressNodeFirewallICMPRule define ingress node firewall rule for ICMP and ICMPv6 protocols.
+	// +optional
+	ICMPRule *IngressNodeFirewallICMPRule `json:"icmpRule,omitempty"`
+
+	// Protocol can be ICMP, ICMPv6, TCP, SCTP or UDP.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum="icmp";"icmpv6";"tcp";"udp";"sctp"
+	Protocol IngressNodeFirewallRuleProtocolType `json:"protocol"`
+
+	// Action can be allow or deny, default action is deny
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum="allow";"deny"
+	// +optional
+	Action IngressNodeFirewallActionType `json:"action,omitempty"`
+}
+
+// ProtocolType defines the protocol types that are supported
+type IngressNodeFirewallRuleProtocolType string
+
+const (
+	// ProtocolTypeICMP refers to the ICMP protocol
+	ProtocolTypeICMP IngressNodeFirewallRuleProtocolType = "icmp"
+
+	// ProtocolTypeICMPv6 refers to the ICMPv6 protocol
+	ProtocolTypeICMPv6 IngressNodeFirewallRuleProtocolType = "icmpv6"
+
+	// ProtocolTypeTCP refers to the TCP protocol, for either IPv4 or IPv6
+	ProtocolTypeTCP IngressNodeFirewallRuleProtocolType = "tcp"
+
+	// ProtocolTypeUDP refers to the UDP protocol, for either IPv4 or IPv6
+	ProtocolTypeUDP IngressNodeFirewallRuleProtocolType = "udp"
+
+	// ProtocolTypeSCTP refers to the SCTP protocol, for either IPv4 or IPv6
+	ProtocolTypeSCTP IngressNodeFirewallRuleProtocolType = "sctp"
+)
+
+// IngressNodeFirewallActionType indicates whether an IngressNodeFirewallRule allows or denies traffic
+// +kubebuilder:validation:Pattern=`^allow|deny$`
+type IngressNodeFirewallActionType string
+
+const (
+	IngressNodeFirewallAllow IngressNodeFirewallActionType = "allow"
+	IngressNodeFirewallDeny  IngressNodeFirewallActionType = "deny"
+)
+
+// IngressNodeFirewallRules define ingress node firewall rule
+type IngressNodeFirewallRules struct {
+	// SourceCIDRS is A list of CIDRs from which we apply node firewall rule
+	// +kubebuilder:validation:one-of[]=Type=string;Format:cidr
+	SourceCIDRs []string `json:"sourceCIDRs"`
+	// FirewallProtocolRules is A list of per protocol ingress node firewall rules
+	FirewallProtocolRules []IngressNodeFirewallProtocolRule `json:"rules,omitempty"`
```

Comment:
Must the list have one entry per protocol? In which case this will want to be a map type list (`listType=map`) with a list map key (`listMapKey=protocol`)

Reply:
no u can have multiple tcp entries for example
```yaml
 - order: 20
      protocol: tcp
      protoRule:
        ports: 800
      action: deny
 - order: 25
     protocol: tcp
      protoRule:
        ports: 900
      action: allow
```

Reply:
The order must be unique though? In which case the map key is order